src/cargo/ops/cargo_compile.rs

```diff
@@ -155,6 +155,11 @@ pub fn compile_pkg<'a>(root_package: &Package,
         s.split(' ')
     }).map(|s| s.to_string()).collect::<Vec<String>>();
 
+    if spec.len() > 0 && (no_default_features || features.len() > 0) {
+        bail!("features cannot be modified when the main package \
+               is not being built");
+    }
```

Comment:
Ah this was removed in 3332c91f983af579d56f36d6413ec0758122def8, so perhaps a rebase brought it back in by accident?